runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkStateInternals.java

```diff
@@ -419,7 +429,130 @@ public int hashCode() {
     }
   }
 
-  private static class FlinkBagState<K, T> implements BagState<T> {
+  private static class FlinkOrderedListState<T> implements OrderedListState<T> {
+    private final StateNamespace namespace;
+    // TODO: Consider instead storing MapStateDescriptor<Instant, T> - at least for some state
+    // backends this will
+    // be stored sorted, which could save time sorting.
+    private final ListStateDescriptor<TimestampedValue<T>> flinkStateDescriptor;
+    private final KeyedStateBackend<ByteBuffer> flinkStateBackend;
+
+    FlinkOrderedListState(
+        KeyedStateBackend<ByteBuffer> flinkStateBackend,
+        String stateId,
+        StateNamespace namespace,
+        Coder<T> coder,
+        SerializablePipelineOptions pipelineOptions) {
+      this.namespace = namespace;
+      this.flinkStateBackend = flinkStateBackend;
+      this.flinkStateDescriptor =
+          new ListStateDescriptor<>(
+              stateId, new CoderTypeSerializer<>(TimestampedValueCoder.of(coder), pipelineOptions));
+    }
+
+    @Override
+    public Iterable<TimestampedValue<T>> readRange(Instant minTimestamp, Instant limitTimestamp) {
+      return readAsMap().subMap(minTimestamp, limitTimestamp).values();
```

Comment:
OT: I've always missed a more sophisticated lifecycle for these states. Something that would allow us to cache the map and invalidate it when it's no longer needed (eg. when we're processing the next key).